h2o-core/src/main/java/hex/Model.java

```diff
@@ -2223,24 +2224,56 @@ public boolean testJavaScoring(Frame data, Frame model_predictions, EasyPredictM
     }
   }
 
+  static <T extends Lockable<T>> int deleteAll(Key<T>[] keys) {
+    int c = 0;
+    for (Key k : keys) {
+      T t = DKV.getGet(k);
+      if (t != null) {
+        t.delete(); //delete all subparts
+        c++;
+      }
+    }
+    return c;
+  }
+
   public void deleteCrossValidationModels() {
     if (_output._cross_validation_models != null) {
-      for (Key k : _output._cross_validation_models) {
-        Model m = DKV.getGet(k);
-        if (m!=null) m.delete(); //delete all subparts
+      Log.info("Cleaning up CV Models for " + this._key.toString());
+      int count = deleteAll(_output._cross_validation_models);
+      Log.info(count+" CV models were removed");
+      try {
+        write_lock();
+        _output._cross_validation_models = null;
+        update();
+      } finally {
+        unlock();
       }
     }
   }
 
   public void deleteCrossValidationPreds() {
+    boolean needsUpdate = false;
     if (_output._cross_validation_predictions != null) {
-      for (Key k : _output._cross_validation_predictions) {
-        Frame f = DKV.getGet(k);
-        if (f!=null) f.delete();
-      }
+      Log.info("Cleaning up CV Predictions for " + this._key.toString());
+      int count = deleteAll(_output._cross_validation_predictions);
+      Log.info(count+" CV predictions were removed");
+      needsUpdate = true;
     }
+
     if (_output._cross_validation_holdout_predictions_frame_id != null) {
       _output._cross_validation_holdout_predictions_frame_id.remove();
+      needsUpdate = true;
+    }
+
+    if (needsUpdate) {
+      try {
+        write_lock();
+        _output._cross_validation_predictions = null;
+        _output._cross_validation_holdout_predictions_frame_id = null;
+        update();
+      } finally {
+        unlock();
+      }
     }
```

Comment:
something in the likes of the following snippet may be useful in Lockable class:
```
<T extends Lockable<T>> void safeUpdate(Function<T> updateFn) {
    if (this.is_wlocked()) {
      updateFn.apply(this);
    } else {
      try {
        this.write_lock();
        updateFn.apply(this);
        this.update();
      } finally {
        this.unlock();
      }
    }
  }
```